### PR TITLE
[Xtensa] Implement branch analysis.

### PR DESCRIPTION
**Pull request by andreisfr**

### CODE REVIEW

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 1]
```diff
@@ -185,3 +187,479 @@ void XtensaInstrInfo::loadImmediate(MachineBasicBlock &MBB,
...
+    return MI.getOperand(2).getMBB();
+
+  case Xtensa::BEQI:
+  case Xtensa::BNEI:
+  case Xtensa::BLTI:
+  case Xtensa::BLTUI:
+  case Xtensa::BGEI:
+  case Xtensa::BGEUI:
+    return MI.getOperand(2).getMBB();
+
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:288]: Unnecessary empty lines

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 2]
```diff
@@ -185,3 +187,479 @@ void XtensaInstrInfo::loadImmediate(MachineBasicBlock &MBB,
...
+  default:
+    llvm_unreachable("Unknown branch opcode");
+  }
+}
+
+bool XtensaInstrInfo::analyzeBranch(MachineBasicBlock &MBB,
+                                    MachineBasicBlock *&TBB,
+                                    MachineBasicBlock *&FBB,
+                                    SmallVectorImpl<MachineOperand> &Cond,
+                                    bool AllowModify = false) const {
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:339]: This family of functions should be implemented independently of branch relaxation. They're more used for branch folding

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 3]
```diff
@@ -185,3 +187,479 @@ void XtensaInstrInfo::loadImmediate(MachineBasicBlock &MBB,
...
+    report_fatal_error(
+        "Branch offsets outside of the signed 32-bit range not supported");
+  XtensaConstantPoolValue *C =
+      XtensaConstantPoolMBB::Create(MF->getFunction().getContext(), &DestBB, 0);
+  unsigned Idx = ConstantPool->getConstantPoolIndex(C, Align(4));
+
+  // FIXME: A virtual register must be used initially, as the register
+  // scavenger won't work with empty blocks (SIInstrInfo::insertIndirectBranch
+  // uses the same workaround).
+  Register ScratchReg = MRI.createVirtualRegister(&Xtensa::ARRegClass);
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:491]: I think this got fixed actually. The workaround might still be there though

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp (hunk 2 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:339]: Thank you for comment. Is it ok if I will correct this PR and remove branch relaxation functionality? The PR then will contain only branch analysis implementation.

Comment (arsenm) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:339]: sure

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp (hunk 1 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp]: Currently this code is removed as redundant.

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp (hunk 3 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp]: Currently this code is removed as redundant.

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp (hunk 2 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:254]: Corrected

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.h [hunk 4]
```diff
@@ -74,6 +74,37 @@ class XtensaInstrInfo : public XtensaGenInstrInfo {
...
+
+  unsigned removeBranch(MachineBasicBlock &MBB,
+                        int *BytesRemoved = nullptr) const override;
+
+  unsigned insertBranch(MachineBasicBlock &MBB, MachineBasicBlock *TBB,
+                        MachineBasicBlock *FBB, ArrayRef<MachineOperand> Cond,
+                        const DebugLoc &DL,
+                        int *BytesAdded = nullptr) const override;
+
+  unsigned InsertBranchAtInst(MachineBasicBlock &MBB,
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.h:93]: Start with lowercase

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 5]
```diff
@@ -185,3 +185,301 @@ void XtensaInstrInfo::loadImmediate(MachineBasicBlock &MBB,
...
+  case Xtensa::BLTI:
+    Cond[0].setImm(Xtensa::BGEI);
+    return false;
+  case Xtensa::BGEUI:
+    Cond[0].setImm(Xtensa::BLTUI);
+    return false;
+  case Xtensa::BLTUI:
+    Cond[0].setImm(Xtensa::BGEUI);
+    return false;
+
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:231]: Inconsistent blank lines here and a few other places

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 6]
```diff
@@ -185,3 +185,301 @@ void XtensaInstrInfo::loadImmediate(MachineBasicBlock &MBB,
     report_fatal_error("Unsupported load immediate value");
   }
 }
+
+bool XtensaInstrInfo::reverseBranchCondition(
+    SmallVectorImpl<MachineOperand> &Cond) const {
+  assert(Cond.size() <= 4 && "Invalid branch condition!");
+
+  switch (Cond[0].getImm()) {
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:193]: Not sure all of these are tested?

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.h (hunk 4 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.h]: Corrected

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 7]
```diff
@@ -185,3 +185,301 @@ void XtensaInstrInfo::loadImmediate(MachineBasicBlock &MBB,
...
+                                    bool AllowModify = false) const {
+  // Most of the code and comments here are boilerplate.
+
+  // Start from the bottom of the block and work up, examining the
+  // terminator instructions.
+  MachineBasicBlock::iterator I = MBB.end();
+  while (I != MBB.begin()) {
+    --I;
+    if (I->isDebugValue())
+      continue;
```
[the remark this answers is not in the document]
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:260]: I prepared new test which checks analyzeBranch and reverseBranchCondition functionality. I hope I understand your comment correctly.

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp (hunk 6 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:193]: I tested most of cases for reverseBranchCondition and it seems it works correctly.

**Review by arsenm — APPROVED** [review]

**Pull request merged**